frame/treasury/src/benchmarking.rs

```diff
@@ -0,0 +1,211 @@
+	let _ = T::Currency::make_free_balance_be(&caller, value);
+	let reason = vec![0; length as usize];
+	let awesome_person = account("awesome", 0, SEED);
+	(caller, reason, awesome_person)
+}
+
+// Create the pre-requisite information needed to call `tip_new`.
+fn setup_tip<T: Trait>(r: u32, t: u32) ->
+	Result<(T::AccountId, Vec<u8>, T::AccountId, BalanceOf<T>), &'static str>
+{
+	for i in 0 .. t {
+		let member = account("member", i, SEED);
+		T::Tippers::add(&member);
+	}
+
+	ensure!(T::Tippers::count() == t as usize, "problem creating tippers");
+	let caller = account("member", t - 1, SEED);
+	let reason = vec![0; r as usize];
+	let beneficiary = account("beneficiary", t, SEED);
+	let value = T::Currency::minimum_balance().saturating_mul(100.into());
+	Ok((caller, reason, beneficiary, value))
+}
+
+// Create `t` new types for the tip proposal with `hash`.
+// This function automatically moves forward the block number to a time which
+// would resolve the tipping process.
+fn create_tips<T: Trait>(t: u32, hash: T::Hash, value: BalanceOf<T>) -> Result<(), &'static str> {
+	for i in 0 .. t {
+		let caller = account("member", i, SEED);
+		ensure!(T::Tippers::contains(&caller), "caller is not a tipper");
+		Treasury::<T>::tip(RawOrigin::Signed(caller).into(), hash, value)?;
+	}
+	frame_system::Module::<T>::set_block_number(T::TipCountdown::get() * 10.into());
+	Ok(())
+}
+
+// Create proposals that are approved for use in `on_finalize`.
+fn create_approved_proposals<T: Trait>(n: u32) -> Result<(), &'static str> {
+	for i in 0 .. n {
+		let (caller, value, lookup) = setup_proposal::<T>(i);
+		Treasury::<T>::propose_spend(
+			RawOrigin::Signed(caller).into(),
+			value,
+			lookup
+		)?;
+		let proposal_id = ProposalCount::get() - 1;
+		Treasury::<T>::approve_proposal(RawOrigin::Root.into(), proposal_id)?;
+	}
+	ensure!(Approvals::get().len() == n as usize, "Not all approved");
+	Ok(())
+}
+
+const MAX_BYTES: u32 = 16384;
+const MAX_TIPPERS: u32 = 100;
+
+benchmarks! {
+	_ { }
+
+	propose_spend {
+		let u in 0 .. 1000;
+		let (caller, value, beneficiary_lookup) = setup_proposal::<T>(u);
+	}: _(RawOrigin::Signed(caller), value, beneficiary_lookup)
+
+	reject_proposal {
+		let u in 0 .. 1000;
+		let (caller, value, beneficiary_lookup) = setup_proposal::<T>(u);
+		Treasury::<T>::propose_spend(
+			RawOrigin::Signed(caller).into(),
+			value,
+			beneficiary_lookup
+		)?;
+		let proposal_id = ProposalCount::get() - 1;
+	}: _(RawOrigin::Root, proposal_id)
+
+	approve_proposal {
+		let u in 0 .. 1000;
+		let (caller, value, beneficiary_lookup) = setup_proposal::<T>(u);
+		Treasury::<T>::propose_spend(
+			RawOrigin::Signed(caller).into(),
+			value,
+			beneficiary_lookup
+		)?;
+		let proposal_id = ProposalCount::get() - 1;
+	}: _(RawOrigin::Root, proposal_id)
+
+	report_awesome {
+		let r in 0 .. MAX_BYTES;
+		let (caller, reason, awesome_person) = setup_awesome::<T>(r);
+	}: _(RawOrigin::Signed(caller), reason, awesome_person)
```

Comment:
This also depends on how much Tips and Reasons storage are full. I guess we should make the assumption that all those storage: Proposal, Reasons, Tips, have all negligible performance variation wether they are full or not because they will not hold that much value anyway.

Reply:
See my other comments